lib/internal/fs/promises.js

```diff
@@ -663,19 +666,46 @@ async function writeFile(path, data, options) {
   options = getOptions(options, { encoding: 'utf8', mode: 0o666, flag: 'w' });
   const flag = options.flag || 'w';
 
-  if (!isArrayBufferView(data)) {
-    validateStringAfterArrayBufferView(data, 'data');
-    data = Buffer.from(data, options.encoding || 'utf8');
+  if (isStream(data) || isIterable(data)) {
+    const fd = await open(path, flag, options.mode);
+    if (options.signal?.aborted) {
+      throw lazyDOMException('The operation was aborted', 'AbortError');
+    }
+    try {
+      for await (const buf of data) {
+        await fd.write(buf);
+      }
+    } finally {
+      await fd.close();
+    }
+  } else {
+    if (!isArrayBufferView(data)) {
+      validateStringAfterArrayBufferView(data, 'data');
+      data = Buffer.from(data, options.encoding || 'utf8');
+    }
+
+    if (path instanceof FileHandle) {
+      return writeFileHandle(path, data, options.signal);
+    }
+
+    const fd = await open(path, flag, options.mode);
+    if (options.signal?.aborted) {
+      throw lazyDOMException('The operation was aborted', 'AbortError');
+    }
+    return PromisePrototypeFinally(writeFileHandle(fd, data), fd.close);
   }
+}
 
-  if (path instanceof FileHandle)
-    return writeFileHandle(path, data, options.signal);
+function isStream(obj) {
+  return !!(obj && typeof obj.pipe === 'function');
+}
 
-  const fd = await open(path, flag, options.mode);
-  if (options.signal?.aborted) {
-    throw lazyDOMException('The operation was aborted', 'AbortError');
+function isIterable(obj) {
+  if (!obj || typeof obj === 'string' || obj.buffer || ArrayIsArray(obj)) {
+    return false;
   }
-  return PromisePrototypeFinally(writeFileHandle(fd, data), fd.close);
+  return typeof obj[SymbolIterator] === 'function' ||
+    typeof obj[SymbolAsyncIterator] === 'function';
```

Comment:
You should use `isTypedArray` from `internal/util/types`. Also, you don't need to test for string&array when `Symbol.asyncIterator` is there:
```suggestion
  return SymbolAsyncIterator in obj || (
    SymbolIterator in obj &&
    typeof obj !== 'string' &&
    !ArrayIsArray(obj) &&
    !isTypedArray(obj));
```